### ISSUE TITLE
Should unions including `Any` trigger `ANN401` (or some other rule)?

### ISSUE THREAD

**Issue by bersbersbers** [astral-sh/ruff#5458]
```python
from typing import Any

def fun_ann401(abc: Any) -> None:
    print(abc)

def fun_no_error(abc: Any | None = None) -> None:
    print(abc)
```

gives

> bug.py:4:21: ANN401 Dynamically typed expressions (typing.Any) are disallowed in `abc`

I wonder why that error is not also emitted for `fun_no_error`. `Any | None` is at least as undefined as `Any` is.

Another option would be to add a new rule that says
> `Any | Foo` is redundant and should be replaced by `Any`.

**Comment by charliermarsh** [comment]
I think it should be included, yeah.

**Comment by charliermarsh** [comment]
Maybe we can reuse the logic that @dhruvmanila added for implicit optionals.

**Comment by dhruvmanila** [comment]
> Maybe we can reuse the logic that @dhruvmanila added for implicit optionals.

If I understand correctly, do you mean to check if there's `typing.Any` instead of `None` in the `TypingTarget`? Or, do you just mean to use the `|` union iterator to check if there's `Any` in it?

**Comment by charliermarsh** [comment]
The latter! Does that make sense?

**Comment by dhruvmanila** [comment]
It does but then `Union[Any, None]` is same as `Any | None` which won't be triggered.

**Comment by dhruvmanila** [comment]
Actually, this then brings up the question if we should look for `Any` in other places as well (`Optional[Any]`).

I believe the main question would be "does this type resolve to `Any`"?

**Comment by charliermarsh** [comment]
Err, sorry, I may have misunderstood your question. We're just looking to understand whether the type resolves to `Any`, I think.

**Comment by dhruvmanila** [comment]
Right, then I believe the following cases should also be considered although the original [plugin doesn't support them](https://github.com/sco1/flake8-annotations#dynamic-typing-caveats).
* `Any | ...`
* `Union[Any, ...]`
* `Optional[Any]`
* `Annotated[<any of the above variants>, ...]`
* Forward references?

Does this make sense?

I don't think `Any` in container types (e.g., `list[Any]`) should be considered, right?

**Comment by charliermarsh** [comment]
Yes, agreed.

**Comment by dhruvmanila** [comment]
It should be simple enough to extract out the logic from `implicit-optional` check and add another method to check for `Any`.
